fix: Handle copy of folders containing live photos

artonge opened the pull request:
We need to recursively look for live photos in the folder, and then handle them as usual.

Fix https://github.com/nextcloud/server/issues/49289
Fix https://github.com/nextcloud/server/issues/49307

- Need https://github.com/nextcloud-libraries/nextcloud-cypress/pull/692

kesselb commented on apps/files/lib/Listener/SyncLivePhotosListener.php:
```diff
@@ -193,4 +184,34 @@ private function handleDeletion(BeforeNodeDeletedEvent $event, Node $peerFile):
+	private function handleCopyRecursive(Event $event, Node $sourceNode, Node $targetNode): void {
+		if ($sourceNode instanceof Folder && $targetNode instanceof Folder) {
+			foreach ($sourceNode->getDirectoryListing() as $sourceChild) {
+				$targetNode = $targetNode->get($sourceChild->getName());
+				$this->handleCopyRecursive($event, $sourceChild, $targetNode);
```
```suggestion
				$targetChild = $targetNode->get($sourceChild->getName());
				$this->handleCopyRecursive($event, $sourceChild, $targetChild);
```

Otherwise, it fails for the second file.

kesselb commented:
getDirectoryListing may return a listing like

- a.jpg
- a.mov
- b.jpg
- b.mov

Loop:

1. handleCopy for a.jpg will copy a.mov too. 
2. handleCopy for a.mov will then copy a.jpg again.

It would be nice to skip the copy process for a.mov and a.jpg for the second execution if that's possible without making the whole logic much more complex.

artonge committed:
fix: Correctly create NonExistingFolder during copy

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
fix: Handle copy of folders containing live photos

We need to recursively look for live photos in the folder,
and then handle them as usual.

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
test: Copying a folder containing live photos

Signed-off-by: Louis Chemineau <louis@chmn.me>

come-nc commented on apps/dav/lib/Connector/Sabre/Directory.php:
```diff
@@ -447,7 +447,10 @@ public function copyInto($targetName, $sourcePath, INode $sourceNode) {
-				return $this->fileView->copy($sourcePath, $destinationPath);
+				$copyOkay = $this->fileView->copy($sourcePath, $destinationPath);
+				if (!$copyOkay) {
+					throw new \Sabre\DAV\Exception\Forbidden('');
```
```suggestion
					throw new \Sabre\DAV\Exception('');
```
Why Forbidden?

kesselb commented:
Regression: It's no longer possible to copy a folder in the current folder. 

**master:**

[Screencast From 2024-11-21 19-25-26.webm](https://github.com/user-attachments/assets/1cd75707-e29f-4edc-b3f9-b366aea8b220)

**this pr:**

[Screencast From 2024-11-21 19-26-00.webm](https://github.com/user-attachments/assets/fcf69680-2a51-4c8d-9ab0-e4b83946b735)

kesselb commented:
Test cases:

| Scenario | Expected | Result |
|--------|--------|--------|
| Copy live photo "a.jpg" to folder where "a.mov" exists  | Copy not possible | ✅  |
| Copy live photo "a.jpg" to folder where "a.mov" exists (batch action)  | Copy not possible | ✅ |
| Move live photo "a.jpg" to folder where "a.mov" exists  | Move not possible | ✅  |
| Move live photo "a.jpg" to folder where "a.mov" exists (batch action)  | Move not possible | ✅ |
| Copy a folder with live photos (same storage) | Copy not possible | ❌ |
| Copy a folder with live photos (same storage, batch action)| Copy not possible | ❌ |
| Copy a folder with live photos (external local to primary s3) | Copy not possible | ❌ |
| Copy a folder with live photos (external local to primary s3, batch action)| Copy not possible | ❌ |
| Move a folder with live photos (same storage) | Live photos work after move | ✅ |
| Move a folder with live photos (same storage, batch action)| Live photos work after move | ✅ |
| Move a folder with live photos (external local to primary s3) | Live photos work after move | ✅ |
| Move a folder with live photos (external local to primary s3, batch action)| Live photos work after move | ✅ |

artonge committed:
fix: Copying a live photo to a folder with a conflicting mov file

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
fix: Throw exception when copy failed

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
/backport to stable30

kesselb commented:
https://github.com/nextcloud/server/pull/49293#issuecomment-2491972877 is not resolved. The current behavior is a bit different. There's an error copying the folder, but the folder and files are copied, but the metadata is missing. 

Expected outcome: Metadata copied, Success Toast. 

[Screencast From 2024-12-04 14-09-11.webm](https://github.com/user-attachments/assets/eb2069ce-423f-4be4-a079-5b18d230dab8)

kesselb commented:
I've updated the test results table: https://github.com/nextcloud/server/pull/49293#issuecomment-2492001753

Currently, the .jpg file is copied, but the process is aborted upon detecting the existing .mov file. Please verify if this new behavior is intentional. I’m fine with keeping it as is, though the previous version was preferable as it avoided leaving behind stale files from incomplete copy attempts.

kesselb commented:
I've tested "Copying a folder with Live Photos" using desktop client 3.15 (no vfs) with `bulkupload.enabled` true and false: The metadata is not copied, the live photo state is lost.

| bulkupload.enabled | requests |
|--------|--------|
| true | mkcol + post for the bulk entpoint |
| false | mkcol + put for each file in the folder |

artonge commented:
@kesselb I was not able to reproduce all of your failed scenario. Can you make sure that you pulled the latest version of the PR?

Regarding the Desktop client behavior, please see with a desktop engineer.

kesselb commented:
In addition to our clients, I had a brief look with Nautilus using the Gnome virtual file system.

Nautilus/GVfs is sending a mkcol request followed by a copy request for each file in the folder. 

[Screencast From 2024-12-04 15-37-10.webm](https://github.com/user-attachments/assets/65023250-18e6-4c46-a81b-8a54a7488dbb)


1. The copy requests for a.jpg led to an internal server error with the following stack trace:

```json
{
  "reqId": "7Q76AXYdcHz8ON5rYQji",
  "level": 3,
  "time": "2024-12-04T14:37:16+00:00",
  "remoteAddr": "172.18.0.1",
  "user": "admin",
  "app": "remote",
  "method": "COPY",
  "url": "/remote.php/dav/files/admin/LivePhotos/a.jpg",
  "message": "OCA\\Files\\Listener\\SyncLivePhotosListener::handleCopy(): Argument #3 ($peerFile) must be of type OCP\\Files\\File, OC\\Files\\Node\\Folder given, called in /var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php on line 240",
  "userAgent": "gvfs/1.56.0",
  "version": "31.0.0.6",
  "exception": {
    "Exception": "TypeError",
    "Message": "OCA\\Files\\Listener\\SyncLivePhotosListener::handleCopy(): Argument #3 ($peerFile) must be of type OCP\\Files\\File, OC\\Files\\Node\\Folder given, called in /var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php on line 240",
    "Code": 0,
    "Trace": [
      {
        "file": "/var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php",
        "line": 240,
        "function": "handleCopy",
        "class": "OCA\\Files\\Listener\\SyncLivePhotosListener",
        "type": "->",
        "args": [
          {
            "__class__": "OC\\Files\\Node\\File"
          },
          {
            "__class__": "OC\\Files\\Node\\File"
          },
          {
            "__class__": "OC\\Files\\Node\\Folder"
          }
        ]
      },
      {
        "file": "/var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php",
        "line": 56,
        "function": "handleCopyRecursive",
        "class": "OCA\\Files\\Listener\\SyncLivePhotosListener",
        "type": "->",
        "args": [
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          },
          {
            "__class__": "OC\\Files\\Node\\File"
          },
          {
            "__class__": "OC\\Files\\Node\\File"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/EventDispatcher/ServiceEventListener.php",
        "line": 68,
        "function": "handle",
        "class": "OCA\\Files\\Listener\\SyncLivePhotosListener",
        "type": "->",
        "args": [
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          }
        ]
      },
      {
        "file": "/var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php",
        "line": 220,
        "function": "__invoke",
        "class": "OC\\EventDispatcher\\ServiceEventListener",
        "type": "->",
        "args": [
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          },
          "OCP\\Files\\Events\\Node\\NodeCopiedEvent",
          {
            "__class__": "Symfony\\Component\\EventDispatcher\\EventDispatcher"
          }
        ]
      },
      {
        "file": "/var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php",
        "line": 56,
        "function": "callListeners",
        "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
        "type": "->",
        "args": [
          [
            {
              "__class__": "Closure"
            },
            {
              "__class__": "Closure"
            },
            {
              "__class__": "Closure"
            },
            {
              "__class__": "Closure"
            }
          ],
          "OCP\\Files\\Events\\Node\\NodeCopiedEvent",
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/EventDispatcher/EventDispatcher.php",
        "line": 67,
        "function": "dispatch",
        "class": "Symfony\\Component\\EventDispatcher\\EventDispatcher",
        "type": "->",
        "args": [
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          },
          "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
        ]
      },
      {
        "file": "/var/www/html/lib/private/EventDispatcher/EventDispatcher.php",
        "line": 79,
        "function": "dispatch",
        "class": "OC\\EventDispatcher\\EventDispatcher",
        "type": "->",
        "args": [
          "OCP\\Files\\Events\\Node\\NodeCopiedEvent",
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/Files/Node/HookConnector.php",
        "line": 194,
        "function": "dispatchTyped",
        "class": "OC\\EventDispatcher\\EventDispatcher",
        "type": "->",
        "args": [
          {
            "__class__": "OCP\\Files\\Events\\Node\\NodeCopiedEvent"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/legacy/OC_Hook.php",
        "line": 85,
        "function": "postCopy",
        "class": "OC\\Files\\Node\\HookConnector",
        "type": "->",
        "args": [
          {
            "oldpath": "/LivePhotos/a.jpg",
            "newpath": "/LivePhotos (Copy)/a.jpg"
          }
        ]
      },
      {
        "file": "/var/www/html/lib/private/Files/View.php",
        "line": 907,
        "function": "emit",
        "class": "OC_Hook",
        "type": "::",
        "args": [
          "OC_Filesystem",
          "post_copy",
          {
            "oldpath": "/LivePhotos/a.jpg",
            "newpath": "/LivePhotos (Copy)/a.jpg"
          }
        ]
      },
      {
        "file": "/var/www/html/apps/dav/lib/Connector/Sabre/Directory.php",
        "line": 450,
        "function": "copy",
        "class": "OC\\Files\\View",
        "type": "->",
        "args": [
          "/LivePhotos/a.jpg",
          "/LivePhotos (Copy)/a.jpg"
        ]
      },
      {
        "file": "/var/www/html/3rdparty/sabre/dav/lib/DAV/Tree.php",
        "line": 150,
        "function": "copyInto",
        "class": "OCA\\DAV\\Connector\\Sabre\\Directory",
        "type": "->",
        "args": [
          "a.jpg",
          "/LivePhotos/a.jpg",
          {
            "__class__": "OCA\\DAV\\Connector\\Sabre\\File"
          }
        ]
      },
      {
        "file": "/var/www/html/3rdparty/sabre/dav/lib/DAV/CorePlugin.php",
        "line": 667,
        "function": "copy",
        "class": "Sabre\\DAV\\Tree",
        "type": "->",
        "args": [
          "files/admin/LivePhotos/a.jpg",
          "files/admin/LivePhotos (Copy)/a.jpg"
        ]
      },
      {
        "file": "/var/www/html/3rdparty/sabre/event/lib/WildcardEmitterTrait.php",
        "line": 89,
        "function": "httpCopy",
        "class": "Sabre\\DAV\\CorePlugin",
        "type": "->",
        "args": [
          {
            "__class__": "Sabre\\HTTP\\Request"
          },
          {
            "__class__": "Sabre\\HTTP\\Response"
          }
        ]
      },
      {
        "file": "/var/www/html/3rdparty/sabre/dav/lib/DAV/Server.php",
        "line": 472,
        "function": "emit",
        "class": "Sabre\\DAV\\Server",
        "type": "->",
        "args": [
          "method:COPY",
          [
            {
              "__class__": "Sabre\\HTTP\\Request"
            },
            {
              "__class__": "Sabre\\HTTP\\Response"
            }
          ]
        ]
      },
      {
        "file": "/var/www/html/apps/dav/lib/Connector/Sabre/Server.php",
        "line": 48,
        "function": "invokeMethod",
        "class": "Sabre\\DAV\\Server",
        "type": "->",
        "args": [
          {
            "__class__": "Sabre\\HTTP\\Request"
          },
          {
            "__class__": "Sabre\\HTTP\\Response"
          }
        ]
      },
      {
        "file": "/var/www/html/apps/dav/lib/Server.php",
        "line": 395,
        "function": "start",
        "class": "OCA\\DAV\\Connector\\Sabre\\Server",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/html/apps/dav/appinfo/v2/remote.php",
        "line": 21,
        "function": "exec",
        "class": "OCA\\DAV\\Server",
        "type": "->",
        "args": []
      },
      {
        "file": "/var/www/html/remote.php",
        "line": 145,
        "args": [
          "/var/www/html/apps/dav/appinfo/v2/remote.php"
        ],
        "function": "require_once"
      }
    ],
    "File": "/var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php",
    "Line": 149,
    "message": "OCA\\Files\\Listener\\SyncLivePhotosListener::handleCopy(): Argument #3 ($peerFile) must be of type OCP\\Files\\File, OC\\Files\\Node\\Folder given, called in /var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php on line 240",
    "exception": {},
    "CustomMessage": "OCA\\Files\\Listener\\SyncLivePhotosListener::handleCopy(): Argument #3 ($peerFile) must be of type OCP\\Files\\File, OC\\Files\\Node\\Folder given, called in /var/www/html/apps/files/lib/Listener/SyncLivePhotosListener.php on line 240"
  }
}
```

2. The forbidden requests seem to confuse Nautilus/GVfs. I guess we need a different handling here with individual copy requests and return okay if the mov file was part of a LivePhoto and already moved with the jpg.

kesselb commented:
When deleting a folder with live photos, the peer files are removed with the source file. That seems to be confusing for Nautilus/GVfs as well because some nodes are not found. I guess Nautilus/GVfs could ignore missing files if we are going to delete the collection anyway. 

[Screencast From 2024-12-04 15-38-30.webm](https://github.com/user-attachments/assets/aa23922d-cbfc-4295-9795-f93dcf6a6256)

kesselb commented:
> > @kesselb I was not able to reproduce all of your failed scenario. Can you make sure that you pulled the latest version of the PR?
> 
> ```
> gh pr checkout 49293 -f
> M	3rdparty
> Already on 'artonge/fix/handle_folders_copy_live_photos'
> Your branch is up to date with 'origin/artonge/fix/handle_folders_copy_live_photos'.
> HEAD is now at bafc1e5245b fix: Throw exception when copy failed
> ```

Sorry, forgot to mention: I merged master locally.

kesselb commented:
The peer id for a.jpg was wrong, and therefore the copy operations failed. 
It now works as expected. 

---

The situation with Nautilus is not ideal. Copying a folder sends a mkcol request and then a copy request for each file. The user is then asked if a.mov should be replaced because a.mov was already copied with a.jpg. Similar case for deleting a folder with live photos, a delete request is sent for each child and Nautilus show's a warning that a.mov is not found.

skjnldsv commented:
No 29 backport :thinking: ?
